### PR TITLE
Refator/vm

### PR DESCRIPTION
**Pull request by SuperFola**

### CODE REVIEW

**Comment by rakista112** [comment]
Windows and Mac builds are failing
https://github.com/ArkScript-lang/Ark/runs/4211737902?check_suite_focus=true#step:9:78
`D:\a\Ark\Ark\include\Ark/VM/ExecutionHandler.hpp(20,63): warning C4003: not enough arguments for function-like macro invocation 'max' [D:\a\Ark\Ark\build\lib\modules\console\console.vcxproj]
D:\a\Ark\Ark\include\Ark/VM/ExecutionHandler.hpp(20,63): error C2589: '(': illegal token on right side of '::' [D:\a\Ark\Ark\build\lib\modules\console\console.vcxproj]
D:\a\Ark\Ark\include\Ark/VM/ExecutionHandler.hpp(20): error C2062: type 'unknown-type' unexpected [D:\a\Ark\Ark\build\lib\modules\console\console.vcxproj]
D:\a\Ark\Ark\include\Ark/VM/ExecutionHandler.hpp(20,30): error C2059: syntax error: ')' [D:\a\Ark\Ark\build\lib\modules\console\console.vcxproj]`
Missing compiler directives maybe? idk
Works fine on the other compilers...

File: src/arkreactor/VM/Scope.cpp [hunk 1]
```diff
@@ -2,22 +2,19 @@
...
-
 namespace Ark::internal
 {
     Scope::Scope() noexcept
     {}
 
     void Scope::push_back(uint16_t id, Value&& val) noexcept
     {
-        push_pair(std::move(id), std::move(val));
+        m_data.emplace_back(std::pair<uint16_t, Value>(std::move(id), std::move(val)));
```
Comment (Loki-Astari) [inline src/arkreactor/VM/Scope.cpp:12]: Don't need `std::pair` here.
The emplace_back handles that for you.

Just use: `m_data.emplace_back(std::move(id), std::move(val));`

File: src/arkreactor/VM/Scope.cpp [hunk 2]
```diff
@@ -2,22 +2,19 @@
...
     }
 
     void Scope::push_back(uint16_t id, const Value& val) noexcept
     {
-        push_pair(id, val);
+        m_data.emplace_back(std::pair<uint16_t, Value>(id, val));
```
Comment (Loki-Astari) [inline src/arkreactor/VM/Scope.cpp:17]: Same comment as above.

**Comment by Loki-Astari** [comment]
This is a breaking change in the API.

I would keep the old API (just not implemented).
Slap a `[[depricated]]` marker with an explanation on how to upgrade to the new API.

**Comment by SuperFola** [comment]
@rakista112 ah yes, windows being a jerk as always

That's because termcolor includes windows.h which defines a min & max macro, I will try to fix that

**Comment by SuperFola** [comment]
> This is a breaking change in the API.

Woops, yes. I've never known how to handle this kind of "regression" (leaving aside incrementing version number). Will do the deprecated, that's a nice idea.

**Comment by SuperFola** [comment]
Alright, the REPL isn't starting because of a stack overflow.

This is most likely due to the VM using an execution context allocated on the stack, and said context is very huge. The idea is to create those contexts at runtime, but it bring the following problem: we can multiple contexts, which one should we use? This is solved in a parallel approach, but we'll need to send to each function of the VM which context it should use.

**Comment by SuperFola** [comment]
We need more tuning, even tho this works, because this brings a huge drop in term of performances.

On the same test file, we are going from 27.13 (dev) to 13.33 MIPS (this branch).

**Comment by SuperFola** [comment]
I figured I can't ask for your review because you're not in a team @Loki-Astari. I updated the code as mentioned, anything to add?

**Pull request merged**